Issues importing plugins and TS module augmentation

### Expected behavior

I'm expecting that importing the plugins, registering them and then configure them in chart options, the project is compiled and working 

### Current behavior

I see some issues TS module augmentation used to define the plugins. 
The first issue was related to PR https://github.com/chartjs/Chart.js/pull/11244 which is forcing the plugins to add the default type to avoid issues on TS compiling.

```ts
interface PluginOptionsByType<TType extends ChartType = ChartType>
```

This update has been applied to annotation plugin but there is some open points about this change: https://github.com/chartjs/chartjs-plugin-datalabels/pull/376

This issue has been introduced by version 4.1.0, probably by PR https://github.com/chartjs/Chart.js/pull/10963

Anyway, adding the default type, the compiler is complaining about the plugin options:

```ts
Object literal may only specify known properties, and 'annotation' does not exist in type '_DeepPartialObject<PluginOptionsByType<keyof ChartTypeRegistry, keyof ChartTypeRegistry>>'.
```

See issue in annotation plugin: https://github.com/chartjs/chartjs-plugin-annotation/issues/886

The issue is not related to annotation itself because testing only zoom and datalabels, the issue remains.
It seems that the module augmentation is ignoring the update to the types after the first one, therefore it depends on the order of the imports.




### Reproducible sample

Project published here: https://github.com/chartjs/chartjs-plugin-annotation/issues/854#issuecomment-1514910268

### Optional extra steps/info to reproduce

I changed chart.js version, using 4.3.0 and annotation 3.0.0. Furthermore I have changed index.d.ts of chartjs-zoom-plugin setting the default type.
Furthermore add annotation and zoom (as empty object) to plugins chart options.

### Possible solution

_No response_

### Context

_No response_

### chart.js version

v4.3.0

### Browser name and version

_No response_

### Link to your project

_No response_

I have created a simple repo which can reproduce this issue.

https://github.com/stockiNail/chartjs-11288

Commands:

```javascript
npm i
npm test
```

Console and issue:

```console
> test
> node node_modules/typescript/bin/tsc index.ts

index.ts:21:7 - error TS2322: Type '{ annotation: {}; zoom: {}; }' is not assignable to type '_DeepPartialObject<PluginOptionsByType<"bar", keyof ChartTypeRegistry>>'.
  Object literal may only specify known properties, and 'annotation' does not exist in type '_DeepPartialObject<PluginOptionsByType<"bar", keyof ChartTypeRegistry>>'.

21       annotation: { },
         ~~~~~~~~~~~~~~~

  node_modules/chart.js/dist/types/index.d.ts:2915:3
    2915   plugins: PluginOptionsByType<TType>;
           ~~~~~~~
    The expected type comes from property 'plugins' which is declared here on type '_DeepPartialObject<CoreChartOptions<"bar"> & ElementChartOptions<"bar"> & PluginChartOptions<"bar"> & DatasetChartOptions<"bar"> & ScaleChartOptions<...> & BarControllerChartOptions>'


Found 1 error in index.ts:21
```

Let me recap.

From version 4.0.1 (where the module augmentation was working) to 4.1.0, the rollup config was changed removing `rollup-plugin-dts`.
Having a look to the output files, I saw that the d.ts produced by rollup plugin exported the plugin definitions by an explicit export (no by re-export).
From 4.1.0, the definition are re-exported by `export * from` (in types.ts which exports `export * from './index.js';`) and this is breaking the module augmentation.

Therefore adding adding `PluginOptionsByType` to `src/types.ts` explicit export (row 10), the issue is **SOLVED**.

This issue was also causing the other TS definition problem (@simonbrunel this is related to the PR https://github.com/chartjs/chartjs-plugin-datalabels/pull/376 that I have submitted in datalabels and where you correctly shared your doubts):

```console
Error: node_modules/chartjs-plugin-annotation/types/index.d.ts:6:13 - error TS2428: 
  All declarations of 'PluginOptionsByType' must have identical type parameters.6   
  interface PluginOptionsByType<TType extends ChartType> {
```

which was fixed (as workaround) adding the default type (see https://github.com/chartjs/Chart.js/pull/11244, which should be reverted I think).

But now, I think there is another decision to take. I can submit a PR to fix this issue but there are other definitions that plugins/controller could augment and are not in a explicit export.
For instance, the zoom plugin is augmenting `UpdateMode` which is not explicitly exported. That means if other plugin will augment as well, adding another update mode, this issue will occur again for  `UpdateMode` (I have tested it, failing, in the repo that I created for this issue). Another example is `ElementOptionsByType` (augmented by annotation plugin).

@etimberg @kurkle @simonbrunel @dangreen what do you think? I think the best thing is to solve this issue as proposed above, and the other defintions which can be augment, we can take time for a solution. Make sense?

--- a/docs/developers/plugins.md
+++ b/docs/developers/plugins.md
@@ -194,7 +194,7 @@ For example, to provide typings for the [`canvas backgroundColor plugin`](../con
 import {ChartType, Plugin} from 'chart.js';
 
 declare module 'chart.js' {
-  interface PluginOptionsByType<TType extends ChartType = ChartType> {
+  interface PluginOptionsByType<TType extends ChartType> {
     customCanvasBackgroundColor?: {
       color?: string
     }

--- a/src/types.ts
+++ b/src/types.ts
@@ -45,6 +45,10 @@ export {
   RadialLinearScale,
   TimeScale,
   TimeSeriesScale,
+  PluginOptionsByType,
+  ElementOptionsByType,
+  ChartDatasetProperties,
+  UpdateModeEnum,
   registerables
 } from './types/index.js';
 export * from './types/index.js';

